### DISCUSSION TITLE
MelonDS android local multiplayer test

### DISCUSSION THREAD

**Discussion by aimdloop** [rafaelvcaetano/melonDS-android#66] [Ideas]
PSteps-
1-first i tried to use two phones bt the emulator was giving communication error every time

![Screenshot_20210216-024706187](https://user-images.githubusercontent.com/47573399/108047900-8d0cb380-7070-11eb-8ab1-113a14943186.jpg)
2-then i download this app to clone the melonds app on the same phone.
https://play.google.com/store/apps/details?id=com.lbe.parallel.intl
3-Then i used two emulator by split the screen.Its very hard to control the emulator using on screen controller when in spilt screen.Bt i managed to do that.
3-Then i tried to establish a connection using ds download play and i succeed.If the layout becomes fixed when emulators in split screen then i guess it is possible to establish a local multiplayer connection.
![Screenshot_2021-02-16-14-47-52-73_e7877f9796c3bfb5a5339f6b3513bed1](https://user-images.githubusercontent.com/47573399/108039381-38643b00-7066-11eb-8eed-aeb09ae5d27a.jpg)

**Comment by rafaelvcaetano** [comment]
Man, this is really awesome! I will address the layout problem.

**Reply by YuvrajSingh0099** [reply]
Can I also join the project? I don't have any experience in game development.

**Comment by ProPaint** [comment]
Maybe you could add a way to run to instances of the emulator like it's done with some other emu as myboy to make things easier, I guess it would need more powerful hardware though.

**Reply by rafaelvcaetano** [reply]
The problem is not the device's power but rather the way melonDS was designed that really makes that difficult. To change that, would mean changing almost all its code. Otherwise, it would be relatively simple.

**Comment by ghost** [comment]
Umm I dunno... I can't run the rom when using the app

**Reply by aimdloop** [reply]
Open a issu in github with detailed information.

**Comment by rafaelvcaetano** [comment]
Let me know what you can achieve with the latest release!

**Reply by aimdloop** [reply]
Now It's easy to establish a local multiplayer connection.Bt it sometimes give communication error.While testing I transfered 6 pokemon from pokemon soul silver(gen 4) to pokemon white 2(gen 5) using ds download play and after the transfer this 2 games was saving.Bt suddenly my pokemon soul silver save file was corrupted.I will sent a screen record as soon as possible.

![Screenshot_2021-02-26-07-37-13-40_e7877f9796c3bfb5a5339f6b3513bed1](https://user-images.githubusercontent.com/47573399/109251387-97b80d00-7815-11eb-99ad-9042e0b24a5b.jpg)
![Screenshot_2021-02-26-07-27-29-65_e7877f9796c3bfb5a5339f6b3513bed1](https://user-images.githubusercontent.com/47573399/109251468-c209ca80-7815-11eb-8584-8924d4c35a2c.jpg)
![Screenshot_2021-02-26-07-27-35-15_e7877f9796c3bfb5a5339f6b3513bed1](https://user-images.githubusercontent.com/47573399/109251478-c7671500-7815-11eb-9352-a86708821a21.jpg)
![Screenshot_2021-02-26-07-34-01-54_e7877f9796c3bfb5a5339f6b3513bed1](https://user-images.githubusercontent.com/47573399/109251493-cafa9c00-7815-11eb-8df2-65bb78e4efcb.jpg)
![Screenshot_2021-02-26-07-28-07-87_e7877f9796c3bfb5a5339f6b3513bed1](https://user-images.githubusercontent.com/47573399/109251506-cfbf5000-7815-11eb-9ea2-b02c138aef61.jpg)

**Comment by aimdloop** [comment]
In realme 5i I can't go further than that it continueously loading.But in redmi 5 plus I can go little further than that and that was the last step of local communication by saving the game.Bt unknown reason I always get communication error in that step.And if I try to follow this process again in redmi 5+ I get this save file corrupt error.Idk what's the cause.Maybe my save file has a problem.

![Screenshot_20210226-130146_melonDS](https://user-images.githubusercontent.com/47573399/109268560-71a26500-7835-11eb-8086-9c6644bcc069.png)

**Reply by rafaelvcaetano** [reply]
I think this happens because writing save files is currently very slow since it causes a lot of writes to the storage. The PC version of melonDS has a new feature that holds the save file in memory for some time and then writes everything at once to the storage. That should help solve this problem.

**Reply by aimdloop** [reply]
Maybe that will solve the problem.I will also try to change my save file.

**Reply by aimdloop** [reply]
I can record the log using matlag while testing the multiplayer. Maybe you can check what's going on.But I really don't know it will be beneficial as I am a noob and maybe it can record others apks log too idk.Bt I will try to close other apks.What u say?

**Reply by rafaelvcaetano** [reply]
Well, that's not really going to help. The emulator is performing as expected (as in, it doesn't crash of freeze) and so, there's no useful information about its state that can be extracted.

**Reply by aimdloop** [reply]
I tried using other save files bt the condition is same.

**Comment by aimdloop** [comment]
Pictochat is working fine.Bt sometimes the msg don't come.

**Comment by aimdloop** [comment]
Trading pokemon is possible i just knew.In pokemon games some pokemons only evolve while trading.Since to trade needs local multiplayer connection so no one could evolve their pokemon in nds games using android.Bt in melonds it is possible rn using same tricks.

Battle with urself is also possible rn.Bt i think no one will battle with himself.Ppl battle with friends and families.So they will use two phones.Thus the emulator give communication error while using 2 phones so noone will battle one same phn.

**Comment by ProPaint** [comment]
It's also working with Mario Kart (not really useful tho)
![Screenshot_20210317-212619_melonDS](https://user-images.githubusercontent.com/30475603/111536733-ebf64300-876a-11eb-8d32-95999020da04.png)

**Reply by aimdloop** [reply]
I tested also.Needs two phones to play this kind of game.But it is not possible rn.

**Comment by aimdloop** [comment]
In this video i wasn't able to go further from certain point.But in this latest beta [v1.4.0](https://github.com/rafaelvcaetano/melonDS-android/releases/tag/beta-1.4.0) i can fully complete the process.Man it's working perfectly.The game is also saving.I can fully use the ds download play in my realme 5i.Idk how.Probably "Merge latest melonDS changes (mostly for the support for asynchronous game saving feature)" for this update.I haven't tested it on my other phone.I will test it later.

If i try to screen record the emulator the emulators become very laggy.Thats why the game don't save.If i turn off the screen recorder the games save.There is no way to screen record this.It needs more high performance phones to record.I tried 480p 2 bitrate.Bt still same.I think increasing emulator performance will solve all local multiplayer related issu.For now i can give a screenshot-
![Screenshot_2021-03-30-04-19-55-22_e7877f9796c3bfb5a5339f6b3513bed1](https://user-images.githubusercontent.com/47573399/112907777-3a62f480-9110-11eb-8713-ce76ad5623fc.jpg)
![Screenshot_2021-03-30-04-22-20-30_e7877f9796c3bfb5a5339f6b3513bed1](https://user-images.githubusercontent.com/47573399/112907789-3e8f1200-9110-11eb-8d97-32d76c7e966f.jpg)
![Screenshot_2021-03-30-04-22-29-24_e7877f9796c3bfb5a5339f6b3513bed1](https://user-images.githubusercontent.com/47573399/112907796-418a0280-9110-11eb-8ff3-c6f24bd0aae6.jpg)

**Reply by Redtextrainer** [reply]
Hello @shabbirmahmud hope you can see this reply, can you please let me know how you did this? I was able to transfer just once, do you have any type of guide on how to do it? Thanks

**Reply by aimdloop** [reply]
@Redtextrainer I had recording on how i did it but unfortunately my youtube account got suspended for some reason. I think u can do it if you read this discussion or you can join some discord or telegram groups and ask people about it.

**Comment by aimdloop** [comment]
Those who are facing communication error while locally trade.
Turn off filtering. Turn on randomized mac address on both emulators. Turn on fast forward after booting the game. Idk what made the difference. Bt it is working for some users who was getting communication error.

**Reply by Chen-jj** [reply]
Awesome, it works!

![image](https://user-images.githubusercontent.com/11807297/180989758-a3fc97b9-20f4-4256-a7ab-0a7fc7759dad.png)

**Reply by alan-david-223** [reply]
> 

Have you tried doing this on two separated devices recently or is it pointless?

I've been trying to play pokemon with other people using two Android devices, we manage to connect to the trade room, but it's really unstable and disconnects as soon as we interact.

PS: all the videos you posted are unavailable :(

**Reply by Chen-jj** [reply]
@alan-david-223 I tried both two Android Phones and Android Phone with MacBook before I found this discussion.But both failed.

The concrete way I known is using two separate MelonDS App on either Windows PC or Android Phone with [parallel App](https://play.google.com/store/apps/details?id=com.lbe.parallel.intl).

Maybe you could copy the `.sav` file of yourself as well as your friend's to one device for trading.

Last but not least, closing **C-GEAR** before trading would be more stable.

**Comment by DaveG95** [comment]
I'm trying to use local multiplayer in the game Animal Crossing Wild World to visit my main city with a secondary save through MelonDS on android, but when I'm connecting and the message "there's a visitor" appears, the main game closes. Has anyone managed to make this connection through the emulator?

I don't know what I'm doing wrong, I've tried all the settings presented in this conversation and others.

![Screenshot_20220421-172956_melonDS](https://user-images.githubusercontent.com/104177768/164577562-1ed19e84-6a7a-4d29-ac0b-22742d4a4c3c.jpg)

**Comment by GuiLimen** [comment]
i try to use ds download play on pokémon platinum but melon ds copy doesn't find the game can you tell me how to use ds download play or does it only work on 5th gen games?

**Comment by Snowball750** [comment]
That's Really Cool! Wish This Was Implemented In The Future.

As for the problem of save corruption, I think maybe it's because the android port still can't connect to the internet.

**Comment by Drago-Cuven** [comment]
@rafaelvcaetano  do you know if there is any way to do local multiplayer between 2 different phones?
Mainly for games like NSMB.

**Comment by Drago-Cuven** [comment]
Or local play on Pokemon black and white.

**Comment by tempdenisa20-gif** [comment]
> PSteps- 1-first i tried to use two phones bt the emulator was giving communication error every time
> 
> ![Screenshot_20210216-024706187](https://user-images.githubusercontent.com/47573399/108047900-8d0cb380-7070-11eb-8ab1-113a14943186.jpg) 2-then i download this app to clone the melonds app on the same phone. https://play.google.com/store/apps/details?id=com.lbe.parallel.intl 3-Then i used two emulator by split the screen.Its very hard to control the emulator using on screen controller when in spilt screen.Bt i managed to do that. 3-Then i tried to establish a connection using ds download play and i succeed.If the layout becomes fixed when emulators in split screen then i guess it is possible to establish a local multiplayer connection. ![Screenshot_2021-02-16-14-47-52-73_e7877f9796c3bfb5a5339f6b3513bed1](https://user-images.githubusercontent.com/47573399/108039381-38643b00-7066-11eb-8eed-aeb09ae5d27a.jpg)

Hey I know it's unlikely that you are going to respond but, did you have different firmwares and bios or the same on both melon DSS?

**Comment by mickydraper5** [comment]
Anybody had any luck recently with getting local Multiplayer to work?
